builtin/logical/pki/path_acme_revoke.go

```diff
@@ -0,0 +1,180 @@
+		if !ok {
+			return nil, fmt.Errorf("invalid type (%T; expected string) for field 'certificate': %w", rawCertificate, ErrMalformed)
+		}
+
+		certBytes, err := base64.RawURLEncoding.DecodeString(certBase64)
+		if err != nil {
+			return nil, fmt.Errorf("failed to base64 decode certificate: %v: %w", err, ErrMalformed)
+		}
+
+		cert, err = x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse certificate: %v: %w", err, ErrMalformed)
+		}
+	} else {
+		return nil, fmt.Errorf("bad request was lacking required field 'certificate': %w", ErrMalformed)
+	}
+
+	rawReason, present := data["reason"]
+	if present {
+		reason, ok := rawReason.(float64)
+		if !ok {
+			return nil, fmt.Errorf("invalid type (%T; expected float64) for field 'reason': %w", rawReason, ErrMalformed)
+		}
+
+		if int(reason) != 0 {
+			return nil, fmt.Errorf("Vault does not support revocation reasons (got %v; expected omitted or 0/unspecified): %w", int(reason), ErrBadRevocationReason)
+		}
+	}
+
+	// If the certificate expired, there's no point in revoking it.
+	if cert.NotAfter.Before(time.Now()) {
+		return nil, fmt.Errorf("refusing to revoke expired certificate: %w", ErrMalformed)
+	}
+
+	// Fetch the CRL config as we need it to ultimately do the
+	// revocation. This should be cached and thus relatively fast.
+	config, err := b.crlBuilder.getConfigWithUpdate(acmeCtx.sc)
+	if err != nil {
+		return nil, fmt.Errorf("unable to revoke certificate: failed reading revocation config: %v: %w", err, ErrServerInternal)
+	}
+
+	// Load our certificate from storage to ensure it exists and matches
+	// what was given to us.
+	serial := serialFromCert(cert)
+	certEntry, err := fetchCertBySerial(acmeCtx.sc, "certs/", serial)
+	if err != nil {
+		return nil, fmt.Errorf("unable to revoke certificate: err reading global cert entry: %v: %w", err, ErrServerInternal)
+	}
+	if certEntry == nil {
+		return nil, fmt.Errorf("unable to revoke certificate: no global cert entry found: %w", ErrServerInternal)
+	}
+
+	// Validate that the provided certificate matches the stored
+	// certificate. This completes the chain of:
+	//
+	//     provided_auth -> provided_cert == stored cert.
+	//
+	// Allowing revocation to be safe.
+	//
+	// We use the non-subtle unsafe bytes equality check here as we have
+	// already fetched this certificate from storage, thus already leaking
+	// timing information that this cert exists. The user could thus simply
+	// fetch the cert from Vault matching this serial number via the unauthed
+	// pki/certs/:serial API endpoint.
+	if !bytes.Equal(certEntry.Value, cert.Raw) {
+		return nil, fmt.Errorf("unable to revoke certificate: supplied certificate does not match CA's stored value: %w", ErrMalformed)
+	}
+
+	// Check if it was already revoked; in this case, we do not need to
+	// revoke it again and want to respond with an appropriate error message.
+	revEntry, err := fetchCertBySerial(acmeCtx.sc, "revoked/", serial)
```

Comment:
We should be considering unified revocation paths as well as they can return us the same error about already being revoked I believe.

Reply:
You're right that this technically races if someone did a BYOC revocation of this cert on a separate CRL with unified CRL enabled, but this at worst sets different revocation times for the same cert entry and so doesn't really matter in practice. Here, we'd eventually run the periodic func (which syncs these BYOC'd revocations back to the local storage as revocations), so _eventually_ we'd get `ErrAlreadyRevoked`, but this is a presumably hopefully small window of outage time. :-)